File: mqtt-client-Android/app/src/main/java/io/emqx/mqtt/PublishRecyclerViewAdapter.java
package io.emqx.mqtt;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class PublishRecyclerViewAdapter extends RecyclerView.Adapter<PublishRecyclerViewAdapter.ViewHolder> {

    private final List<Publish> mValues;

    public PublishRecyclerViewAdapter(List<Publish> items) {
        mValues = items;
    }

    @Override
    public PublishRecyclerViewAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_publish, parent, false);
        return new PublishRecyclerViewAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final PublishRecyclerViewAdapter.ViewHolder holder, int position) {
        Publish publish = mValues.get(position);
        holder.mItem = publish;
        holder.mTopicView.setText("Topic: " + publish.getTopic() + "  Qos: " + publish.getQos() + "  Retained: " + publish.isRetained());
        holder.mPayloadView.setText("Payload: " + publish.getPayload());
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView mTopicView;
        public final TextView mPayloadView;
        public Publish mItem;

        public ViewHolder(View view) {
            super(view);
            mView = view;
            mTopicView = view.findViewById(R.id.textView);
            mPayloadView = view.findViewById(R.id.payload);
        }

    }
}

File: mqtt-client-Android-Kotlin/app/src/main/java/io/emqx/mqtt/PublishRecyclerViewAdapter.kt
package io.emqx.mqtt

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class PublishRecyclerViewAdapter(private val mValues: List<Publish>) :
    RecyclerView.Adapter<PublishRecyclerViewAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_publish, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val publish = mValues[position]
        holder.mItem = publish
        holder.mTopicView.text =
            "Topic: " + publish.topic + "  Qos: " + publish.qos + "  Retained: " + publish.isRetained
        holder.mPayloadView.text = "Payload: " + publish.payload
    }

    override fun getItemCount(): Int {
        return mValues.size
    }

    inner class ViewHolder(mView: View) : RecyclerView.ViewHolder(
        mView
    ) {
        val mTopicView: TextView = mView.findViewById(R.id.textView)
        val mPayloadView: TextView = mView.findViewById(R.id.payload)
        var mItem: Publish? = null

    }
}